src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -309,6 +309,28 @@ PathTuples SparqlQleverVisitor::visitTypesafe(
   return joinPredicateAndObject(predicate, objectList);
 }
 
+// ____________________________________________________________________________________
+vector<SparqlPrefix> SparqlQleverVisitor::visitTypesafe(
+    SparqlAutomaticParser::PrologueContext* ctx) {
+  vector<SparqlPrefix> decls = visitVector<SparqlPrefix>(ctx->baseDecl());
+  appendVector(decls, visitVector<SparqlPrefix>(ctx->prefixDecl()));
+  return decls;
```

Comment:
I think you should still store the prefixes in the `_prefixMap` member.
This seems redundant for now, but
we will need it later once we use a stateful visitor for the complete queries.

Reply:
Also the HashMap correctly handles the case that a prefix is redefined (this is somewhat of a corenercase for which there is no real use case, but we should either handle it correctly (the last one counts)

Reply:
The `prefixMap()` member was only used for tests. So with this we'd have duplicate datastructures (the `vector<SparqlPrefix>` and `PrefixMap`). We'd may also be converting the data from one structure to another and back again. `vector<SparqlPrefix>` -> `PrefixMap` (to eliminate duplicates) -> `vector<SparqlPrefix>` (for storage in `ParsedQuery`) -> `PrefixMap`(for calling the ANTLR parser) This seems unnecessaryily complicated.

Reply:
I mean the internally stored prefix map is a crucial part of the visitor, that's why we pass it in in the `parseWithAntlr` helper functions, to get proper `IRI` parsing etc.

The `PrefixMap` is the thing that will still be there, once the parser is finished. The `vector<SparqlPrefix>` is only needed at the moment to communicate with the legacy code.

So it would be even cleaner to let the `visit(PrologueContext)` return a copy of the internally stored prefix map, and let the legacy parser perform its conversion to the `SparqlPrefix` legacy type.